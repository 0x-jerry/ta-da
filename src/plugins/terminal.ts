import { TickTickRenderer, TypeItem, TypeItemType } from '../core'

export function createTerminalRenderer(): TickTickRenderer {
  const core: TickTickRenderer = {
    split(str: string) {
      const isTerminalStyleChar = /^(\x1b\[\d+m)+/g

      const items: TypeItem[] = []

      const ctx = {
        cursor: 0,
        get current() {
          return str[ctx.cursor]
        },
        subStr() {
          return str.slice(ctx.cursor)
        },
      }

      while (ctx.cursor < str.length) {
        const subStr = ctx.subStr()
        const [hit] = subStr.match(isTerminalStyleChar) || []

        if (hit) {
          const char = subStr.slice(0, hit.length)
          ctx.cursor += hit.length

          items.push({
            type: TypeItemType.Invisible,
            content: char,
          })
        } else {
          if (/\s/.test(ctx.current)) {
            items.push({
              type: TypeItemType.Space,
              content: ctx.current,
            })
          } else {
            items.push({
              type: TypeItemType.Text,
              content: ctx.current,
            })
          }

          ctx.cursor++
        }
      }

      return items
    },
    render(item: TypeItem) {
      process.stdout.write(item.content)
    },
  }

  return core
}
