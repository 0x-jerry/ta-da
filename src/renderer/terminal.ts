import { type TadaRenderer, type TadaItem, TadaItemType } from '../types'
import { nextChar } from './_utils'

export function createTerminalRenderer() {
  const core: TadaRenderer = {
    split(str: string) {
      const isTerminalStyleChar = /^(\x1b\[\d+m)+/g

      const items: TadaItem[] = []

      const ctx = {
        cursor: 0,
        get current() {
          return nextChar(this.subStr()) || ''
        },
        subStr() {
          return str.slice(ctx.cursor)
        },
      }

      while (ctx.cursor < str.length) {
        const subStr = ctx.subStr()
        const [hit] = subStr.match(isTerminalStyleChar) || []

        if (hit) {
          const char = subStr.slice(0, hit.length)
          ctx.cursor += hit.length

          items.push({
            type: TadaItemType.Invisible,
            content: char,
          })
        } else {
          const currentChar = ctx.current

          if (/\s/.test(currentChar)) {
            items.push({
              type: TadaItemType.Space,
              content: currentChar,
            })
          } else {
            items.push({
              type: TadaItemType.Text,
              content: currentChar,
            })
          }

          ctx.cursor += currentChar.length
        }
      }

      return items
    },
    render(item: TadaItem) {
      process.stdout.write(item.content)
    },
  }

  return core
}
