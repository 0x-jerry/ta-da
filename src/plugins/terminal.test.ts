import { createTerminalRenderer } from './terminal'

describe('terminal renderer', () => {
  it('should split string correctly', () => {
    const r = createTerminalRenderer()
    const items = r.split('hell o \x1b[1m123\x1b[22m')

    expect(items).matchSnapshot()
  })
})
