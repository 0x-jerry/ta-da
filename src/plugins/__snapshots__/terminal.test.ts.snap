// Vitest Snapshot v1

exports[`terminal renderer > should split string correctly 1`] = `
[
  {
    "content": "h",
    "type": "text",
  },
  {
    "content": "e",
    "type": "text",
  },
  {
    "content": "l",
    "type": "text",
  },
  {
    "content": "l",
    "type": "text",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "o",
    "type": "text",
  },
  {
    "content": " ",
    "type": "space",
  },
  {
    "content": "[1m",
    "type": "invisible",
  },
  {
    "content": "1",
    "type": "text",
  },
  {
    "content": "2",
    "type": "text",
  },
  {
    "content": "3",
    "type": "text",
  },
  {
    "content": "[22m",
    "type": "invisible",
  },
]
`;
